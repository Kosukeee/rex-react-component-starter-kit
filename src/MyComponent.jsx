import React from 'react';
import PropTypes from 'prop-types';
import 'rex-core/dist/css/rex-core.css';
import './MyComponent.scss';

export default function MyComponent({ children, onClick, text, className }) {
  return (
    <div
      className={className}
      onClick={onClick}
      role="link"
      tabIndex={0}
      onKeyDown={onClick}
    >
      <h1>{text}</h1>
      {children}
    </div>
  );
}

MyComponent.defaultProps = {
  children: null,
  text: 'Welcome to React',
  className: 'my-component',
  onClick: () => {},
};

MyComponent.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  text: PropTypes.string,
  className: PropTypes.string,
  onClick: PropTypes.func,
};
