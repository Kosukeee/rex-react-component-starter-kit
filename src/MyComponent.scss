.my-component {
  margin: 10px;
}