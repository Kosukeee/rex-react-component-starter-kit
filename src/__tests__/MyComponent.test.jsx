import { shallow } from 'enzyme';
import React from 'react';
import MyComponent from '../MyComponent';

describe('<MyComponent />', () => {
  const params = {
    text: 'Hello ReX React Component Starter Kit!!',
  };
  const makeSubject = (props, children) =>
    shallow(<MyComponent {...props}>{children}</MyComponent>);

  beforeEach(() => {
    jest.resetModules();
  });

  it('should match snapshot', () => {
    const component = makeSubject(params);
    expect(component.html()).toMatchSnapshot();
  });

  it('should render default-props', () => {
    const component = makeSubject();

    // Text
    expect(component.text()).toBe('Welcome to React');
    // className
    expect(component.hasClass('my-component')).toBe(true);
    // click, no side-effect
    const clickedComponent = component.simulate('click');
    expect(clickedComponent).toBe(component);
  });

  it('should render passed props', () => {
    const component = makeSubject({
      text: 'Hello Jest Test!!',
      className: 'my-custom-class',
    });

    // Text
    expect(component.text()).toBe('Hello Jest Test!!');
    // className
    expect(component.hasClass('my-custom-class')).toBe(true);
  });

  it('simulate click event', () => {
    const clickHandler = jest.fn();
    const component = makeSubject({
      onClick: clickHandler,
    });

    component.find('div').simulate('click');
    expect(clickHandler).toHaveBeenCalled();
  });

  it('should render a child', () => {
    const component = makeSubject(params, <div>child</div>);
    expect(component.contains(<div>child</div>)).toBe(true);

    // Same as above:
    const wrapper = shallow(
      <MyComponent>
        <p className="child">I am a child</p>
      </MyComponent>
    );
    expect(
      wrapper.containsAllMatchingElements([
        <p className="child">I am a child</p>,
      ])
    ).toBe(true);
  });

  // MEMO: Now multiple-children is not allowed
  // it('should render children', () => {
  //   const wrapper = shallow(
  //     <MyComponent>
  //       <p className="child">I am a child</p>
  //       <span className="child">I am a child, too</span>
  //     </MyComponent>
  //   );
  //   expect(
  //     wrapper.containsAllMatchingElements([
  //       <p className="child">I am a child</p>,
  //       <span className="child">I am a child, too</span>,
  //     ])
  //   ).toBe(true);
  // });
});
