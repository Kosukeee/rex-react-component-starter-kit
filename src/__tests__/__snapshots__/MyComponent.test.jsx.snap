// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyComponent /> should match snapshot 1`] = `"<div class=\\"my-component\\" role=\\"presentation\\"><h1>Hello ReX React Component Starter Kit!!</h1></div>"`;
